src/glob.zig

```diff
@@ -0,0 +1,2555 @@
+
+                            break :recursion_idx_bump 0;
+                        }
+
+                        const matches = this.matchPatternImpl(&pattern, entry_name);
+                        if (matches) {
+                            if (is_last) {
+                                try this.appendMatchedPath(entry_name, dir_path, in_root);
+                                continue;
+                            }
+                            break :recursion_idx_bump 1;
+                        }
+                        continue;
+                    };
+
+                    const subdir_parts: []const []const u8 = &[_][]const u8{
+                        dir_path[0..dir_path.len],
+                        entry_name,
+                    };
+
+                    const subdir_entry_name = try std.fs.path.join(this.arena.allocator(), subdir_parts);
+
+                    try this.workbuf.append(this.arena.allocator(), WorkItem.new(subdir_entry_name, component_idx + recursion_idx_bump, .directory));
+                },
+                .sym_link => {
+                    if (this.follow_symlinks) {
+                        @panic("TODO");
+                    }
+
+                    // If following symlinks is not enabled, the default behaviour is to use the same matching strategy used for files:
+                    const matches = this.matchPatternFile(entry_name, component_idx, is_last, &pattern, next_pattern);
+                    if (matches) {
+                        try this.appendMatchedPath(entry_name, dir_path, in_root);
+                    }
+                },
+                else => {},
+            }
+        }
+
+        return .{ .result = undefined };
+    }
+
+    /// A file can only match if:
+    /// a) this is the last pattern, or
+    /// b) it matches the next pattern, provided the current
+    ///    pattern is a double wildcard and the  next pattern is
+    ///    not a double wildcard
+    ///
+    /// Examples:
+    /// a -> `src/foo/index.ts` matches
+    /// b -> `src/**/*.ts` (on 2nd pattern) matches
+    fn matchPatternFile(
+        this: *GlobWalker,
+        entry_name: []const u8,
+        component_idx: u32,
+        is_last: bool,
+        pattern: *Component,
+        next_pattern: ?*Component,
+    ) bool {
+        // Handle case a)
+        if (!is_last) return pattern.syntax_hint == .Double and
+            component_idx + 1 == this.patternComponents.items.len - 1 and
+            next_pattern.?.syntax_hint != .Double and
+            this.matchPatternImpl(next_pattern.?, entry_name);
+
+        // Handle case b)
+        return this.matchPatternImpl(pattern, entry_name);
+    }
+
+    fn matchPatternImpl(
+        this: *GlobWalker,
+        pattern_component: *Component,
+        filepath: []const u8,
+    ) bool {
+        // FIXME: handle utf-16?
+        if (!this.dot and GlobWalker.startsWithDot(filepath)) return false;
+        switch (pattern_component.syntax_hint) {
```

Comment:
Small style thing: `switch` can be an expression

```suggestion
return switch (pattern_component.syntax_hint) {
            .Double, .Single => true,
            .WildcardFilepath => if (comptime !isWindows)
                matchWildcardFilepath(this.pattern[pattern_component.start .. pattern_component.start + pattern_component.len], filepath)
            else
                this.matchPatternSlow(pattern_component, filepath),
            .Literal => if (comptime !isWindows)
                matchWildcardLiteral(this.pattern[pattern_component.start .. pattern_component.start + pattern_component.len], filepath)
            else
                this.matchPatternSlow(pattern_component, filepath),
            else => this.matchPatternSlow(pattern_component, filepath),
};
```